packages/react-spinbutton/src/components/SpinButton/useSpinButton.tsx

```diff
@@ -0,0 +1,316 @@
+  SpinButtonBounds,
+} from './SpinButton.types';
+import { calculatePrecision, precisionRound, getBound, clampWhenInRange } from '../../utils/index';
+import { ChevronUp16Regular, ChevronDown16Regular } from '@fluentui/react-icons';
+
+type InternalState = {
+  value: number;
+  spinState: SpinButtonSpinState;
+  spinTime: number;
+  spinDelay: number;
+  previousTextValue?: string;
+};
+
+const DEFAULT_SPIN_DELAY_MS = 150;
+const MIN_SPIN_DELAY_MS = 80;
+const MAX_SPIN_TIME_MS = 1000;
+
+// This is here to give an ease the mouse held down case.
+// Exact easing it to be defined. Once it is we'll likely
+// pull this out into a util function in the SpinButton package.
+const lerp = (start: number, end: number, percent: number): number => start + (end - start) * percent;
+
+/**
+ * Create the state required to render SpinButton.
+ *
+ * The returned state can be modified with hooks such as useSpinButtonStyles_unstable,
+ * before being passed to renderSpinButton_unstable.
+ *
+ * @param props - props from this instance of SpinButton
+ * @param ref - reference to root HTMLElement of SpinButton
+ */
+export const useSpinButton_unstable = (props: SpinButtonProps, ref: React.Ref<HTMLInputElement>): SpinButtonState => {
+  const nativeProps = getPartitionedNativeProps({
+    props,
+    primarySlotTagName: 'input',
+    excludedPropNames: ['onChange', 'size'],
+  });
+
+  const {
+    value,
+    displayValue,
+    defaultValue,
+    min,
+    max,
+    step = 1,
+    stepPage = 1,
+    precision: precisionFromProps,
+    onChange,
+    size = 'medium',
+    appearance = 'outline',
+    root,
+    input,
+    incrementButton,
+    decrementButton,
+    inputType = 'all',
+  } = props;
+
+  const precision = React.useMemo(() => {
+    return precisionFromProps ?? Math.max(calculatePrecision(step), 0);
+  }, [precisionFromProps, step]);
+
+  const [currentValue, setCurrentValue] = useControllableState({
+    state: value,
+    defaultState: defaultValue,
+    initialState: 0,
+  });
+  const [textValue, setTextValue] = React.useState(
+    value !== undefined && displayValue !== undefined ? displayValue : String(currentValue),
+  );
+  const [spinState, setSpinState] = React.useState<SpinButtonSpinState>('rest');
+  const [atBound, setAtBound] = React.useState<SpinButtonBounds>('none');
+
+  const internalState = React.useRef<InternalState>({
+    value: currentValue,
+    spinState,
+    spinTime: 0,
+    spinDelay: DEFAULT_SPIN_DELAY_MS,
+  });
+
+  const state: SpinButtonState = {
+    size,
+    appearance,
+    spinState,
+    atBound,
+
+    components: {
+      root: 'span',
+      input: 'input',
+      incrementButton: 'button',
+      decrementButton: 'button',
+    },
+    root: resolveShorthand(root, {
+      required: true,
+      defaultProps: nativeProps.root,
+    }),
+    input: resolveShorthand(input, {
+      required: true,
+      defaultProps: {
+        ref,
+        autoComplete: 'off',
+        size: input?.size,
```

Comment:
Why is `size` being set on the input itself?  I noticed you put it in `excludedPropNames` for `getPartitionedNativeProps`, but then add it back here

(for reference, I'd expect it to not be set on the input el at all)

Reply:
Good catch! This is left over from when I was attempting to use `react-input`'s `Input` component rather than `input`. It relies on a `size` prop for some of its styling [which is different](https://github.com/microsoft/fluentui/blob/master/packages/react-input/src/components/Input/Input.types.ts#L38-L45) from the built in `input` `size` attribute.